### PR TITLE
QA missing data

### PR DESCRIPTION
**Pull request by jonahkaye**
Ticket: #[1442](https://github.com/metriport/metriport-internal/issues/1442)

### Dependencies

- Upstream: https://github.com/metriport/metriport/pull/1583

### Description

- [x] missing encounter data added (which added missing condition data too)
- [x] missing dosage on immunization
- [x] missing assessment section
- [x] missing meds + observations from required/optional issues resolved

### Other

- We have less Procedure notes in the QA, but only because they are instead inside Diagnostic Reports which for the context makes more conceptual sense
- Missing drug details in MedicationStatement just a reflection of CDA / MedicationStatement references Medication which has more details. 

### Testing

- Local
  - [x] ran full fhir insert

Check each PR.

### Release Plan

- [ ] Merge this

### CODE REVIEW

File: packages/fhir-converter/src/templates/cda/Header.hbs [hunk 1]
```diff
@@ -1,74 +1,76 @@
-{{#with msg.ClinicalDocument}}
-    {{>Resources/Composition.hbs composition=this ID=(generateUUID (toJsonString this))}},
+{{#with msg.ClinicalDocument as |document|}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Header.hbs:1]: systematically removing this and .. from templating system. Very confusing and time and time again source of context related bugs

File: packages/fhir-converter/src/templates/cda/Header.hbs [hunk 2]
```diff
@@ -1,74 +1,76 @@
...
-                {{>Resources/Practitioner.hbs practitioner=../legalAuthenticator.assignedEntity ID=practitionerId.Id}},
+    {{#if document.legalAuthenticator}}
+        {{#with (evaluate 'Utils/GeneratePractitionerId.hbs' obj=document.legalAuthenticator.assignedEntity) as |practitionerId|}}
+            {{>Resources/Practitioner.hbs practitioner=document.legalAuthenticator.assignedEntity ID=practitionerId.Id}},
+        {{/with}}
+    {{/if}}
+
+    {{#if document.componentOf.encompassingEncounter}}
+        {{!-- do we want to include this? Kinda a duplicate of more rich data later in encounter section --}}
+        {{>Resources/Encounter.hbs encounter=document.componentOf.encompassingEncounter ID=(generateUUID (toJsonString document.componentOf.encompassingEncounter))}},
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Header.hbs:15]: this encounter we currently include is kinda like a poor mans version of encounter later on in document

File: packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs [hunk 3]
```diff
@@ -1,10 +1,12 @@
-{{#if (contains (toString (toJsonString msg)) '2.16.840.1.113883.10.20.22.2.22.1')}}
-    {{#with (getFirstCdaSectionsByTemplateId msg '2.16.840.1.113883.10.20.22.2.22.1')}}
-            {{#each (toArray this.[2_16_840_1_113883_10_20_22_2_22_1].entry) as |encounterEntry|}}
-                {{>Resources/Encounter.hbs encounter=encounterEntry.encounter ID=(generateUUID (toJsonString encounterEntry))}},
-                {{#with (evaluate 'Utils/GeneratePatientId.hbs' obj=@metriportPatientId) as |patientId|}}
-                	{{>References/Encounter/subject.hbs ID=(generateUUID (toJsonString encounterEntry)) REF=(concat 'Patient/' patientId.Id)}},
-                {{/with}}
+{{#if (contains (toString (toJsonString msg)) '2.16.840.1.113883.10.20.22.2.22')}}
+    {{#with (getFirstCdaSectionsByTemplateId msg '2.16.840.1.113883.10.20.22.2.22.1' '2.16.840.1.113883.10.20.22.2.22') as |encounters|}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs:2]: We were missing encounters because some were under: '2.16.840.1.113883.10.20.22.2.22' and not at all under '2.16.840.1.113883.10.20.22.2.22.1'

File: packages/fhir-converter/src/templates/cda/Sections/Hospital_Discharge_Medications.hbs [hunk 4]
```diff
@@ -37,7 +37,7 @@
                             {{#if entryRelationship.supply.author.assignedAuthor.representedOrganization}}
                                 {{#with (evaluate 'Utils/GenerateOrganizationId.hbs' obj=entryRelationship.supply.author.assignedAuthor.representedOrganization) as |orgId|}}
                                     {{>Resources/Organization.hbs org=entryRelationship.supply.author.assignedAuthor.representedOrganization ID=orgId.Id}},
-                                    {{>References/MedicationAdministration/performer.actor.hbs ID=(generateUUID (toJsonString medAdm.substanceAdministration)) REF=(concat 'Organization/' OrgId.Id)}},
+                                    {{>References/MedicationAdministration/performer.actor.hbs ID=(generateUUID (toJsonString medAdm.substanceAdministration)) REF=(concat 'Organization/' orgId.Id)}},
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Sections/Hospital_Discharge_Medications.hbs:40]: OrgId doesnt exist. Lowercase does. Bug causing undefined Orgs

File: packages/fhir-converter/src/templates/cda/Sections/Hospital_Discharge_Medications.hbs [hunk 5]
```diff
@@ -57,7 +57,7 @@
                             {{#if entryRelationship.supply.performer.assignedEntity.representedOrganization}}
                                 {{#with (evaluate 'Utils/GenerateOrganizationId.hbs' obj=entryRelationship.supply.performer.assignedEntity.representedOrganization) as |orgId|}}
                                     {{>Resources/Organization.hbs org=entryRelationship.supply.performer.assignedEntity.representedOrganization ID=orgId.Id}},
-                                    {{>References/MedicationDispense/performer.actor.hbs ID=(generateUUID (toJsonString entryRelationship.supply)) REF=(concat 'Organization/' OrgId.Id)}},
+                                    {{>References/MedicationDispense/performer.actor.hbs ID=(generateUUID (toJsonString entryRelationship.supply)) REF=(concat 'Organization/' orgId.Id)}},
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Sections/Hospital_Discharge_Medications.hbs:60]: ditto

File: packages/fhir-converter/src/templates/cda/Sections/Notes.hbs [hunk 6]
```diff
@@ -3,10 +3,12 @@
             {{!-- sections here is already an array so this is not a bug --}}
             {{#each sections}}
                 {{#each (multipleToArray this.[2_16_840_1_113883_10_20_22_2_65].entry this.[1_3_6_1_4_1_19376_1_5_3_1_3_5].entry) as |note|}}
-                    {{>Resources/DiagnosticReport.hbs diagReport=note.act ID=(generateUUID (toJsonString note.act))}},
-                    {{#with (evaluate 'Utils/GeneratePatientId.hbs' obj=@metriportPatientId) as |patientId|}}
-                        {{>References/DiagnosticReport/subject.hbs ID=(generateUUID (toJsonString note.act)) REF=(concat 'Patient/' patientId.Id)}},
-                    {{/with}}
+                    {{#if note.act}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Sections/Notes.hbs:6]: weird bug, but multipleToArray results in an array of all the notes and then one extra array item which is a multipleToArray object? This caused an empty resource to be created. Not sure root issue, but adding if here protects

File: packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs (hunk 3 above)
Comment (Goncharo) [reply packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs:2]: So this is a really important find @jonahkaye - here is the full story:

See this immunization template for reference: https://github.com/metriport/metriport/blob/572208c0c20c90cf80f3594aabbb2fffba68f916/packages/fhir-converter/src/templates/cda/Sections/Immunization.hbs#L1-L2

Also, see list of C-CDA section templates for reference: https://build.fhir.org/ig/HL7/cda-ccda-2.2/artifacts.html#2

Note that it checks the "higher-level" section ID, and then it gets the sections matching 2 different IDs. The "higher-level" ID is the "entries optional" ID, and the other one is the "entries required" ID:
- https://build.fhir.org/ig/HL7/cda-ccda-2.2/StructureDefinition-2.16.840.1.113883.10.20.22.2.2.html
- https://build.fhir.org/ig/HL7/cda-ccda-2.2/StructureDefinition-2.16.840.1.113883.10.20.22.2.2.1.html

Several C-CDA sections, including the Encounters one you uncovered follow this pattern - let's sweep through and make sure there's no other sections missing like this.

Comment (jonahkaye) [reply packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs:2]: So my only concern here was that if we used getFirstCdaTemplate function, is their a chance their are multiple sections, one with the required id and one with the optional id?

We could use getAllCDATemplate sections, but then will we get duplicates for the sections that have both the optional and required id?

File: packages/fhir-converter/src/templates/cda/Header.hbs [hunk 7]
```diff
@@ -1,74 +1,76 @@
...
-        {{#if ../patient.guardian}}
-            {{>Resources/RelatedPerson.hbs relatedPerson=../patient.guardian ID=(generateUUID (toJsonString ../patient.guardian))}},
-            {{>References/RelatedPerson/patient.hbs ID=(generateUUID (toJsonString ../patient.guardian)) REF=(concat 'Patient/' patientId.Id)}},
-        {{/if}}
-    {{/with}}
+    {{#with document.recordTarget.patientRole as |patientRole|}}
+        {{#with (evaluate 'Utils/GeneratePatientId.hbs' obj=@metriportPatientId) as |patientId|}}
+            {{>Resources/Patient.hbs patientRole=patientRole ID=patientId.Id}},
+            {{#if patientRole.patient.guardian}}
+                {{#each (toArray patientRole.patient.guardian) as |guardian|}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Header.hbs:68]: guardian is actually an array.

**Review by jonahkaye — COMMENTED** [review]
```
{
  countPerType: {
    Condition: 17018 => 17580, 
    DiagnosticReport: 30906 => 27553,
    Encounter: 18582 => 19384,
    Location: 17626 => 17861,
    Practitioner: 63145 => 63404,
    RelatedPerson: 12240 => 12260
  },
  total: 412438 => 410963
}
```

All the following are increasing because we were missing ~4% of encounters before 
Encounter ⬆️ 
Condition ⬆️ 
Practitioner ⬆️ 
Location ⬆️ 
Org: ⬆️ 

This is 10% decreasing because we eliminated a bug where a dead Diagnostic Report was created every single time w/ multipleToArray
Diagnostic Report ⬇️

**Review by jonahkaye — COMMENTED** [review]
```
{
  countPerType: {
    Medication: 9892 => 12391
    MedicationRequest: 4975 => 8022
    MedicationStatement: 5585 => 8354
    Observation: 151097 => 165048
    Organization: 32101 => 32171
    Practitioner: 63404 => 66830
  },
  total: 412438 => 436719
}
```

**Review by jonahkaye — COMMENTED** [review]
```
{
  countPerType: {
    Medication: 12391 => 12264
    MedicationAdministration: 5387 => 4889
    MedicationStatement: 8354 => 8205
    Organization: 32171 => 32133
    Practitioner: 66830 => 65437
    ServiceRequest: 1
  },
  total: 436719 => 434513
}
```

File: packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js [hunk 8]
```diff
@@ -16,9 +16,10 @@ const convert = require("convert-units");
 
 
 const PERSONAL_RELATIONSHIP_TYPE_CODE = "2.16.840.1.113883.1.11.19563";
-const decimal_regex = /-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/;
+const decimal_regex = /-?(?:(?:0|[1-9][0-9]*)\.?[0-9]*|\.[0-9]+)(?:[eE][+-]?[0-9]+)?/;
```
Comment (jonahkaye) [inline packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js:19]: we were missing decimals without a leading interger, so 
.5 was no good but 0.5 was. Patch for that

File: packages/utils/src/fhir-converter/scripts/compare_total_resource_counts.sh [hunk 9]
```diff
@@ -22,4 +22,4 @@ if [ ! -f "$FILE2" ]; then
 fi
 
 # Run jsondiffpatch
-jsondiffpatch "$FILE1" "$FILE2"
+jsondiffpatch "$FILE2" "$FILE1"
```
Comment (jonahkaye) [inline packages/utils/src/fhir-converter/scripts/compare_total_resource_counts.sh:25]: for readability, more natural to see old -> new instead of new -> old

**Review by jonahkaye — COMMENTED** [review]
```
{
  countPerType: {
    Medication: 12264 => 12391
    MedicationAdministration: 4889 => 5387
    MedicationStatement: 8205 => 8354
    Organization: 32133 => 32171
  },
  total: 434513 => 435325
}
```

**Review by jonahkaye — COMMENTED** [review]
Inserted 9446 files in 4513782 ms. [1 errors]
```
Resources:  {
  AllergyIntolerance: 736,
  Communication: 789,
  Composition: 9405,
  Condition: 13635,
  Consent: 44,
  Coverage: 558,
  Device: 3830,
  DiagnosticReport: 24940,
  Encounter: 19159,
  FamilyMemberHistory: 725,
  Goal: 45,
  Immunization: 2123,
  MedicationAdministration: 2435,
  MedicationRequest: 5629,
  MedicationStatement: 6946,
  Observation: 94614,
  Procedure: 8915,
  RelatedPerson: 271
}
```
Total resources: 194799

File: packages/fhir-converter/src/templates/cda/Resources/Immunization.hbs [hunk 10]
```diff
@@ -38,6 +38,14 @@
             {{/with}}
         {{/each}}
 
+        {{#if imm.doseQuantity}}
+            "doseQuantity":{{>DataType/SimpleQuantity.hbs quantity=imm.doseQuantity}},
+        {{/if}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/Immunization.hbs:43]: added missing dosage info

**Review by jonahkaye — COMMENTED** [review]
```
{
  countPerType: {
    DiagnosticReport: 27547 => 31800
  },
  total: 435325 => 439578
}
```

File: packages/fhir-converter/src/templates/cda/Resources/DiagnosticReport.hbs [hunk 11]
```diff
@@ -11,6 +11,7 @@
         ],
         "status":{{>ValueSet/DiagnosticReportStatus.hbs code=diagReport.statusCode.code}},
         "code":{{>DataType/CodeableConcept.hbs code=diagReport.code}},
+        "category":{{>DataType/CodeableConcept.hbs code=categoryCode}},
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/DiagnosticReport.hbs:14]: cx noted missing

**Review by jonahkaye — COMMENTED** [review]
After slight changes. still chillinnnn at same count

```
{
  "total": 435325,
  "countPerType": {
    "Communication": 790,
    "Composition": 9445,
    "Condition": 17580,
    "Device": 9207,
    "DiagnosticReport": 27547,
    "Encounter": 19384,
    "Location": 17861,
    "Observation": 165048,
    "Organization": 32171,
    "Patient": 9445,
    "Practitioner": 65437,
    "Procedure": 9603,
    "RelatedPerson": 12260,
    "Medication": 12391,
    "MedicationAdministration": 5387,
    "MedicationRequest": 8022,
    "MedicationStatement": 8354,
    "AllergyIntolerance": 819,
    "Coverage": 744,
    "FamilyMemberHistory": 774,
    "Immunization": 2456,
    "Goal": 547,
    "Consent": 53
  }
}
```

File: packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs (hunk 3 above)
Comment (Goncharo) [reply packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs:2]: assuming this isn't a concern any more since we check for exact match?

File: packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js [hunk 12]
```diff
@@ -601,17 +602,18 @@ module.exports.external = [
...
             let sectionObj = msg.ClinicalDocument.component.structuredBody.component[i].section;
-            if (
-              sectionObj.templateId &&
-              JSON.stringify(sectionObj.templateId).includes(templateIds[t])
-            ) {
+            // Normalize sectionObj.templateId to always be an array
+            let templateIdsArray = Array.isArray(sectionObj.templateId) ? sectionObj.templateId : [sectionObj.templateId];
+            
+            // Check if any templateId object in the normalized array has a root value that matches exactly
+            const hasExactMatch = templateIdsArray.some(templateIdObj => templateIdObj && templateIdObj.root === templateIds[t]);
```
[the remark this answers is not in the document]
Comment (Goncharo) [reply packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js:611]: exact match won't cause us to miss sections that relied on this?

not sure if there's any, but I assume this was like this for a reason

File: packages/fhir-converter/src/templates/cda/Sections/Assessments.hbs [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+    {{#with (getFirstCdaSectionsByTemplateId msg '2.16.840.1.113883.10.20.22.2.8')}}
+        {{#each (toArray 2_16_840_1_113883_10_20_22_2_8) as |assessment|}}
+            {{#if assessment.text}}
+                {{#with (extractAndMapTableData assessment.text) as |assessmentText|}}
+                    {{>References/DiagnosticReport/assessment.hbs diagReport=assessment text=assessmentText ID=(generateUUID (toJsonString assessment))}},
+                {{/with}}
+            {{/if}}
+        {{/each}}
+    {{/with}}
+{{/if}}
```
Comment (Goncharo) [inline packages/fhir-converter/src/templates/cda/Sections/Assessments.hbs:11]: let's leave a big fat TODO here so someone else doesn't come along and use it - until we come up with a good place to put these

**Review by Goncharo — COMMENTED** [review]
generally LGTM - left some comments to consider, will approve after they're resolved

File: packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs (hunk 3 above)
Comment (jonahkaye) [reply packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs:2]: Not a concern because I ran a few scripts to test and didnt find any instances of optionalId and requiredId in the same file in **different** sections. They either are both in section, or only one is.

File: packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js:611]: So it didnt matter before because they never used the optional ids, which are shorter than the required ones. For immunizations
2.16.840.1.113883.10.20.22.2.2   -> optional
2.16.840.1.113883.10.20.22.2.2.1 > required

they didnt have the optional one before, and the full one didnt overlap with any other resources. But the optional one if you use include matches with:

2.16.840.1.113883.10.20.22.2.23 -> MedicationEquipment

which means if we use include we lost immunisations if the first resource we find is a MedicationEquipment.  

Tldr: is that we need exact match since we are using the shorter optional Ids, but we arent missing any data, which makes sense logically and also is shown emprically in test counts

File: packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs (hunk 3 above)
Comment (jonahkaye) [reply packages/fhir-converter/src/templates/cda/Sections/Encounters.hbs:2]: Tests show we arent losing data too

**Review by Goncharo — APPROVED** [review]
:shipit:
